File: 18_ml/problems/00_mnist_infer/mnist_infer.f90
! 本物の MNIST 手書き数字を, 学習済みの2層MLPで認識する (推論=forward)。
!   data/mnist_weights.txt : 学習済みの重み (784->128->10)
!   data/mnist_test.txt    : テスト画像 (28x28=784画素, 0..255) と正解ラベル
! 推論の中身は「行列ベクトル積 + 活性化(ReLU) + argmax」。各画像の推論は独立なので並列化できる。
program mnist_infer
  use omp_lib
  implicit none
  integer :: IN, HID, OUT, NT, IN2, i, hh, oo, k, best, lab
  integer(8) :: correct
  real(8) :: s, bestv, t0, elapsed
  real(8), allocatable :: W1(:,:), b1(:), W2(:,:), b2(:), X(:,:), hidv(:)
  integer, allocatable :: y(:), pix(:)

  ! --- 重みの読み込み ---
  open(10, file="data/mnist_weights.txt", status="old", action="read")
  read(10, *) IN, HID, OUT
  allocate(W1(IN,HID), b1(HID), W2(HID,OUT), b2(OUT))   ! W1(k,hh)=W1[hh][k] の並び
  read(10, *) W1        ! HID*IN 個 (hh ごとに IN 個ずつ) を列優先 W1(:,hh) に読み込む
  read(10, *) b1
  read(10, *) W2        ! OUT*HID 個
  read(10, *) b2
  close(10)

  ! --- テスト画像の読み込み (画素 0..255 -> 0..1) ---
  open(10, file="data/mnist_test.txt", status="old", action="read")
  read(10, *) NT, IN2
  allocate(X(IN,NT), y(NT), pix(IN))
  do i = 1, NT
     read(10, *) pix, lab
     X(:,i) = real(pix, 8) / 255.0d0
     y(i) = lab
  end do
  close(10)

  ! --- 推論 ---
  correct = 0
  t0 = omp_get_wtime()
  ! TODO: 各画像の推論は独立。!$omp parallel do reduction(+:correct) で並列化せよ.
  do i = 1, NT
     allocate(hidv(HID))
     do hh = 1, HID                       ! h = ReLU(W1 x + b1)
        s = b1(hh)
        do k = 1, IN
           s = s + W1(k,hh) * X(k,i)
        end do
        hidv(hh) = max(0.0d0, s)
     end do
     best = 1; bestv = -1d300              ! o = W2 h + b2, argmax
     do oo = 1, OUT
        s = b2(oo)
        do hh = 1, HID
           s = s + W2(hh,oo) * hidv(hh)
        end do
        if (s > bestv) then
           bestv = s; best = oo
        end if
     end do
     if (best - 1 == y(i)) correct = correct + 1   ! クラスは 0..9, best は 1..10
     deallocate(hidv)
  end do
  ! TODO: 上で始めた parallel do 領域を閉じる (!$omp end parallel do).
  elapsed = omp_get_wtime() - t0

  print "(a,i0,a,i0,a,f0.2,a)", &
       "MNIST テスト ", NT, " 枚: 正解 ", correct, " 枚, 正解率 = ", &
       100.0d0 * correct / NT, " %"
  print "(a,f0.3,a)", "elapsed = ", elapsed, " sec"
end program mnist_infer

File: 18_ml/problems/00_mnist_infer/mnist_infer.cpp
#include <cstdio>
#include <cstdlib>
#include <omp.h>

/* 本物の MNIST 手書き数字を, 学習済みの2層MLPで認識する (推論=forward)。
   - data/mnist_weights.txt : 学習済みの重み (784->128->10)
   - data/mnist_test.txt    : テスト画像 (28x28=784画素, 0..255) と正解ラベル
   推論の中身は「行列ベクトル積 + 活性化(ReLU) + argmax」。これまで並列化してきた
   行列計算が, そのまま手書き数字の認識になる。各画像の推論は独立なので並列化できる。 */
int main(int argc, char ** argv) {
  int IN, HID, OUT;

  /* --- 重みの読み込み --- */
  FILE * fw = fopen("data/mnist_weights.txt", "r");
  if (!fw) { printf("data/mnist_weights.txt が開けません\n"); return 1; }
  if (fscanf(fw, "%d %d %d", &IN, &HID, &OUT) != 3) return 1;
  double * W1 = (double *)malloc(sizeof(double) * HID * IN);
  double * b1 = (double *)malloc(sizeof(double) * HID);
  double * W2 = (double *)malloc(sizeof(double) * OUT * HID);
  double * b2 = (double *)malloc(sizeof(double) * OUT);
  for (long k = 0; k < (long)HID * IN; k++)  fscanf(fw, "%lf", &W1[k]);
  for (int k = 0; k < HID; k++)              fscanf(fw, "%lf", &b1[k]);
  for (long k = 0; k < (long)OUT * HID; k++) fscanf(fw, "%lf", &W2[k]);
  for (int k = 0; k < OUT; k++)              fscanf(fw, "%lf", &b2[k]);
  fclose(fw);

  /* --- テスト画像の読み込み (画素 0..255 -> 0..1 に正規化) --- */
  FILE * ft = fopen("data/mnist_test.txt", "r");
  if (!ft) { printf("data/mnist_test.txt が開けません\n"); return 1; }
  int NT, IN2;
  if (fscanf(ft, "%d %d", &NT, &IN2) != 2) return 1;
  double * X = (double *)malloc(sizeof(double) * (long)NT * IN);
  int *    y = (int *)malloc(sizeof(int) * NT);
  for (int i = 0; i < NT; i++) {
    for (int k = 0; k < IN; k++) { int v; fscanf(ft, "%d", &v); X[(long)i*IN+k] = v / 255.0; }
    fscanf(ft, "%d", &y[i]);
  }
  fclose(ft);

  /* --- 推論: 各画像を MLP に通して予測クラス(argmax)を求め, 正解数を数える --- */
  long correct = 0;
  double t0 = omp_get_wtime();
  // TODO: 各画像の推論は独立。#pragma omp parallel for reduction(+:correct) で並列化せよ.
  for (int i = 0; i < NT; i++) {
    double h[1024];                       /* 隠れ層 (HID<=1024 を仮定) */
    const double * x = &X[(long)i * IN];
    for (int hh = 0; hh < HID; hh++) {    /* h = ReLU(W1 x + b1) */
      double s = b1[hh];
      const double * w = &W1[(long)hh * IN];
      for (int k = 0; k < IN; k++) s += w[k] * x[k];
      h[hh] = (s > 0.0) ? s : 0.0;
    }
    int best = 0; double bestv = -1e300;  /* o = W2 h + b2, argmax */
    for (int oo = 0; oo < OUT; oo++) {
      double s = b2[oo];
      const double * w = &W2[(long)oo * HID];
      for (int hh = 0; hh < HID; hh++) s += w[hh] * h[hh];
      if (s > bestv) { bestv = s; best = oo; }
    }
    if (best == y[i]) correct++;
  }
  double elapsed = omp_get_wtime() - t0;

  printf("MNIST テスト %d 枚: 正解 %ld 枚, 正解率 = %.2f%%\n",
         NT, correct, 100.0 * correct / NT);
  printf("elapsed = %.3f sec\n", elapsed);
  free(W1); free(b1); free(W2); free(b2); free(X); free(y);
  return 0;
}
